econml/solutions/causal_analysis/_causal_analysis.py

```diff
@@ -0,0 +1,1044 @@
+                                    props=self._point_props(alpha), n=Xtest.shape[0])
+
+    def _local_causal_effect_dict(self, Xtest, alpha=0.05):
+        """
+        Gets the local feature importance as dictionary
+
+        Dictionary entries for predictions, etc. will be nested lists of shape (n_rows, d_y, sum(d_t))
+
+        Only for serialization purposes to upload to AzureML
+        """
+        return self._dict_summary(self._local_effect_inference(Xtest), props=self._point_props(alpha),
+                                  kind='local')
+
+    def _safe_result_index(self, X, feature_index):
+        assert hasattr(self, "_results"), "This instance has not yet been fitted"
+
+        assert np.ndim(X) == 2 and np.shape(X)[1] == self._d_x, (
+            "Shape of X must be compatible with shape of the fitted X, "
+            f"but got shape {np.shape(X)} instead of (n, {self._d_x})"
+        )
+
+        (numeric_index,) = _get_column_indices(X, [feature_index])
+        results = [res for res in self._results
+                   if res.feature_index == numeric_index]
+
+        assert len(results) != 0, f"The feature index supplied was not fitted"
+        (result,) = results
+        return result
+
+    def whatif(self, X, Xnew, feature_index, y):
+        """
+        Get counterfactual predictions when feature_index is changed to Xnew from its observational counterpart.
+
+        Note that this only applies to regression use cases; for classification what-if analysis is not supported.
+
+        Parameters
+        ----------
+        X: array-like
+            Features
+        Xnew: array-like
+            New values of a single column of X
+        feature_index: int or string
+            The index of the feature being varied to Xnew, either as a numeric index or
+            the string name if the input is a dataframe
+        y: array-like
+            Observed labels or outcome of a predictive model for baseline y values
+
+        Returns
+        -------
+        y_new: InferenceResults
+            The predicted outputs that would have been observed under the counterfactual features
+        """
+
+        assert not self.classification, "What-if analysis cannot be applied to classification tasks"
+
+        assert np.shape(X)[0] == np.shape(Xnew)[0] == np.shape(y)[0], (
+            "X, Xnew, and y must have the same length, but have shapes "
+            f"{np.shape(X)}, {np.shape(Xnew)}, and {np.shape(y)}"
+        )
+
+        assert np.size(feature_index) == 1, f"Only one feature index may be changed, but got {np.size(feature_index)}"
+
+        T0 = _safe_indexing(X, feature_index, axis=1)
+        T1 = Xnew
+        result = self._safe_result_index(X, feature_index)
+        inf = result.estimator.effect_inference(
+            X=result.X_transformer.transform(X), T0=T0, T1=T1)
+
+        # we want to offset the inference object by the baseline estimate of y
+        inf.translate(y)
+
+        return inf
```

Comment:
I think we should be returning a pandas data frame again in the same spirit as we return pandas data frames in all other results. So that the user doesn't need to know our inference results object. 

So we should make alpha part of the input and return a pandas data frame with that alpha

Reply:
We can't return an inference object in the other cases because we've got one-per-feature; here, it's easy to get a data frame out if the user wants to by just getting the summary off of the inference object.

Reply:
I know but they need to know about our inferenceresults object and it's properties. That's ok for now, but I would make it more uniform in api to always return a standard object like a pandas frame or an array

Reply:
Ok, we can change this in the next release.

Reply:
Let's put that in the TODOs